### PR TITLE
non-local handlers

### PR DESCRIPTION
**Pull request by gaborcsardi**
These are only called on jump. Mostly straightforward.

### CODE REVIEW

File: src/cleancall.c [hunk 1]
```diff
@@ -10,23 +10,31 @@ static SEXP callbacks = NULL;
 static void push_callback(SEXP stack) {
   SEXP top = CDR(stack);
 
+  SEXP local_lgl = PROTECT(Rf_allocVector(LGLSXP, 1));
```
Comment (lionel-) [inline src/cleancall.c]: Can you use `Rf_ScalarLogical()` please? this way we spare an alloc.

File: src/cleancall.c [hunk 2]
```diff
@@ -10,23 +10,31 @@ static SEXP callbacks = NULL;
...
   SEXP cb = Rf_cons(Rf_cons(fn_extptr, data_extptr), top);
 
   SETCDR(stack, cb);
 
-  UNPROTECT(2);
+  UNPROTECT(3);
 }
 
+struct data_wrapper {
+  SEXP (*fn)(void *data);
```
Comment (lionel-) [inline src/cleancall.c]: General comment: `*` should stick to the type rather then the binding in our C style, right?

```suggestion
  SEXP (*fn)(void* data);
```

File: src/cleancall.c [hunk 3]
```diff
@@ -35,14 +43,23 @@ static void call_exits(void* data) {
 
     void (*fn)(void*) = (void (*)(void*)) R_ExternalPtrAddrFn(CAR(cb));
     void *data = (void*) EXTPTR_PTR(CDR(cb));
+    int nonlocal = LOGICAL(EXTPTR_TAG(CDR(cb)))[0];
 
     // Check for empty pointer in preallocated callbacks
     if (fn) {
+      if (! nonlocal || ! datawpr->success)
```
Comment (lionel-) [inline src/cleancall.c]:
```suggestion
      if (!nonlocal || !(datawpr->success))
```

Spaces after unary operators is non-standard style I think.

File: src/cleancall.c [hunk 4]
```diff
@@ -10,23 +10,31 @@ static SEXP callbacks = NULL;
...
+};
 
 static void call_exits(void* data) {
   // Remove protecting node. Don't remove the preallocated callback on
   // the top as it might contain a handler when something went wrong.
   SEXP top = CDR(callbacks);
 
   // Restore old stack
-  callbacks = (SEXP) data;
+  struct data_wrapper *datawpr = data;
```
Comment (lionel-) [inline src/cleancall.c]:
```suggestion
  struct data_wrapper* datawpr = data;
```

File: src/cleancall.c [hunk 5]
```diff
@@ -35,14 +43,23 @@ static void call_exits(void* data) {
...
 
     // Check for empty pointer in preallocated callbacks
     if (fn) {
+      if (! nonlocal || ! datawpr->success)
       fn(data);
     }
   }
 }
 
+SEXP r_with_cleanup_context_wrap(void *data) {
```
Comment (lionel-) [inline src/cleancall.c]: Should be static and not namespaced in `r_`

```suggestion
static SEXP with_cleanup_context_wrap(void *data) {
```

File: src/cleancall.c [hunk 6]
```diff
@@ -52,13 +69,16 @@ SEXP r_with_cleanup_context(SEXP (*fn)(void* data), void* data) {
...
+
+  SEXP out = R_ExecWithCleanup(r_with_cleanup_context_wrap, &datawpr,
+			       &call_exits, &datawpr);
 
   UNPROTECT(1);
   return out;
 }
 
-void r_push_cleanup(void (*fn)(void* data), void* data) {
+void r_call_save_handler(void (*fn)(void *data), void* data, int nonlocal) {
```
Comment (lionel-) [inline src/cleancall.c]:
```suggestion
static void call_save_handler(void (*fn)(void *data), void* data, int nonlocal) {
```

File: src/init.c [hunk 7]
```diff
@@ -29,6 +35,10 @@ void R_init_cleancall(DllInfo *dll) {
...
   fns_dot_call = Rf_findVar(Rf_install(".Call"), R_BaseEnv);
 
-  R_RegisterCCallable("cleancall", "r_with_cleanup_context", (DL_FUNC) &r_with_cleanup_context);
-  R_RegisterCCallable("cleancall", "r_push_cleanup", (DL_FUNC) &r_push_cleanup);
+  R_RegisterCCallable("cleancall", "r_with_cleanup_context",
+		      (DL_FUNC) &r_with_cleanup_context);
+  R_RegisterCCallable("cleancall", "r_call_on_exit",
+		      (DL_FUNC) &r_call_on_exit);
+  R_RegisterCCallable("cleancall", "r_call_on_nonlocal_exit",
+		      (DL_FUNC) &r_call_on_nonlocal_exit);
```
Comment (lionel-) [inline src/init.c]: Indentation looks off. Ah it seems you're using tabs. Can you untabify the whole PR please?

File: src/cleancall.c [hunk 8]
```diff
@@ -35,14 +43,23 @@ static void call_exits(void* data) {
...
     // Check for empty pointer in preallocated callbacks
     if (fn) {
+      if (! nonlocal || ! datawpr->success)
       fn(data);
     }
   }
 }
 
+SEXP r_with_cleanup_context_wrap(void *data) {
+  struct data_wrapper *cdata = data;
```
Comment (lionel-) [inline src/cleancall.c]:
```suggestion
  struct data_wrapper* cdata = data;
```

**Review by lionel- — APPROVED** [review]
Mostly style comments. Can you untabify all your changes please? Also commit message should be an imperative sentence that starts with a capital letter.

How about `r_call_on_early_exit()`? This seems more suggestive than nonlocal, shorter, and we avoid the casing looking off (my brain stares at nonlocal and wonders if there should be an underscore).

File: src/cleancall.c (hunk 3 above)
Comment (lionel-) [reply src/cleancall.c]: Maybe `!early_handler || !(state->success)` would be easier to read?

**Comment by gaborcsardi** [comment]
Thanks, looks better! Will merge after CIs done.

**Pull request merged**